File: 921-spiral-matrix-iii/spiral-matrix-iii.js
var spiralMatrixIII = function(rows, cols, rStart, cStart) {
    let size = rows * cols
    let result = []
    let i = rStart
    let j = cStart
    let steps = 1
    let directions = [[0, 1], [1,0], [0, -1], [-1, 0]]
    let k = 0

    function isValid(row, col) {
        if (row < 0 || col < 0 || row >= rows || col >= cols) return false
        return true
    }

    while (result.length < size) {
        for (let a = 0; a < 2; a++) {
            let [x, y] = directions[k]
            for (let step = 0; step < steps; step++) {
                if (isValid(i, j)) {
                    result.push([i, j])
                }
                i = i + x
                j = j + y
            }
            k = (k + 1) % 4
        }
        steps += 1
    }

    return result
};

File: 921-spiral-matrix-iii/spiral-matrix-iii.py
class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> List[List[int]]:
        output = []
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        i = rStart
        j = cStart
        steps = 1
        k = 0

        def isValid(i, j):
            if ((i < 0 or i >= rows) or (j < 0 or j >= cols)):
                return False
            return True
        
        while (len(output) < rows * cols):
                
            for _ in range(2):
                direction = directions[k]
                x, y = direction

                for _ in range(steps):

                    if (isValid(i, j)):
                        output.append([i, j]) 
                    i = x + i
                    j = j + y

                k = (k + 1) % len(directions)
                
            steps += 1

        return output
        